### PR TITLE
Authorization rules

### PR DESCRIPTION
**Pull request by faisalakb**
In this project apply the following features : 

Add a role column to the users table.
A user can delete a post if it is theirs or if they have an admin role (column role has value "admin"). Use CanCanCan for this authorization.

    Add the "Delete" button to the view and make sure that only authorized users can see it.

A user can delete a comment if it is theirs or if they have an admin role (column role has value "admin"). Use CanCanCan for this authorization.

### CODE REVIEW

File: app/controllers/posts_controller.rb [hunk 1]
```diff
@@ -1,36 +1,51 @@
 class PostsController < ApplicationController
   before_action :set_user, only: %i[index show new create]
+  before_action :set_post, only: %i[show destroy]
```
Comment (Ibizugbe) [inline app/controllers/posts_controller.rb:3]: Nice job so far😉😉 Kindly add the `load_and_authorize_resource` method in your post controller. This method is a declarative way of defining both the loading of a resource and authorizing a user to perform certain actions on that resource. When used in the post controller, it automatically loads the appropriate resource and authorizes the requested action.
```suggestion
  before_action :set_post, only: %i[show destroy]
  
  load_and_authorize_resource
```

**Review by Ibizugbe — CHANGES_REQUESTED** [review]
Hi @faisalakb  @Omwomo 👋🏼👋🏼👋🏼, 

require_changes


Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights✅✅✅

- [x] Linters are passing👍🏼.
- [x] `gitflow` is used👍🏼.
- [x] Ability class has been appropriately setup👍🏼.


## Required Changes ♻️

- [ ] You currently haven't added the cancancan gem to your gem file, kindly do well to add them as this gives you the privilege to access the rich features of the library😉.
- [ ] I also noticed your User table hasn't been updated to have the role column which is making your app to break when trying to create a post😉.
image
image

- [ ] Kindly note that it is currently impossible to create and view a post because your `cancancan` library hasn't been correctly setup. Kindly ensure your app works well, it is possible to create a post, and add comments and a user can delete only his posts and comments, an admin can delete everyone's posts and comments. Kindy ensure you read through the requirements and ensure they are all fulfilled before requesting a review, just to avoid blockers🙏🏼😉

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/posts_controller.rb (hunk 1 above)
Comment (faisalakb) [reply app/controllers/posts_controller.rb:3]: Please check i already added the cancancan gem. 
can

Comment (Ibizugbe) [reply app/controllers/posts_controller.rb:3]: Hello @faisalakb  As you can see from the green line on 76 and 77, The changes were made without being pushed to GitHub. This is a picture of your current gemfile which ends at 75.

image

Please do well to commit your changes👍🏼.

**Comment by faisalakb** [comment]
Please check i updated the above comments.

**Comment by faisalakb** [comment]
Please check now i updated above comment and push pending files as well

**Comment by shahadat3669** [comment]
@faisalakb, Well! while I was reviewing your project! I notice that you pushed two new commits. It's best practice to make substantial updates to your code before requesting a review. After the review request, try to avoid pushing new commits until you receive feedback. This helps maintain a clear review process and reduces the need for the reviewer to reevaluate the changes.

**Comment by faisalakb** [comment]
ok

File: tmp/letter_opener/1701447060_3803942_498327e/rich.html [hunk 2]
```diff
@@ -0,0 +1,117 @@
+<html>
+  <head>
+    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
+    
+      <title>Confirmation instructions</title>
+    
+
```
Comment (shahadat3669) [inline tmp/letter_opener/1701447060_3803942_498327e/rich.html]: - [ ] Well done with the code, you updated a lot based on the previous reviewer's request. but I noticed there are multiple extra blank lines scattered throughout your code those are updated after the previous reviewer comments. So I make the request to changes them. It's best practice to keep the code clean and free from unnecessary blank lines. Please consider removing these extra blank lines to improve the code's readability and maintainability. Keep up the good work! 👍🧹📄

**Review by shahadat3669 — CHANGES_REQUESTED** [review]
Hi @faisalakb, @Omwomo , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.




## Don't Worry 🙏

It's okay to have bugs in your code. 🆗 Please see this code review as an opportunity to learn practical coding skills and discover new tips that can improve your code quality 💯 and help you **get even better** 💼.

**I will guide you on how to address these areas down below 👇.**

## Highlights
✅ No linter Errors
✅ Has CanCan Gem in Gemfile : gem 'cancancan'
✅ Has CanCanCan `ability` class that defines the set of rules for user - post interactions.



 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._



Cheers and Happy coding!👏👏👏

## Have any doubt ❓

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

<table  width="100%" align="center">
  <tr>
    <td colspan="4" align="center">
      <strong>Contact me: Shahadat Hossain</strong>
    </td>
  </tr>
  <tr>
    <td align="center">
      <a href="https://github.com/shahadat3669">
        GitHub
      </a>
    </td>
    <td align="center">
      <a href="https://www.linkedin.com/in/shahadat-cseng/">
        LinkedIn
      </a>
    </td>
    <td align="center">
      <a href="https://app.slack.com/client/T47CT8XPG/D05GT4E4WAY/rimeto_profile/U04426SD71V">
        Slack
      </a>
    </td>
    <td align="center">
      <a href="mailto:shahadat3669@gmail.com?subject=Microverse%20Code%20Review">
        Email
      </a>
    </td>
  </tr>
</table>

## ⚠️ WARNING ⚠️

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: tmp/letter_opener/1701447060_3803942_498327e/rich.html (hunk 2 above)
Comment (faisalakb) [reply tmp/letter_opener/1701447060_3803942_498327e/rich.html]: @shahadat3669  could you please inform me in which file i have extra spaces. ?

Comment (shahadat3669) [reply tmp/letter_opener/1701447060_3803942_498327e/rich.html]: @faisalakb check it out.
image

Comment (faisalakb) [reply tmp/letter_opener/1701447060_3803942_498327e/rich.html]: @shahadat3669 these not my files. they are generated by letter opener ..According my understanding it doesn't make any sense. Please explain me why we need to remove any spaces or editing in a file that are generated by some system 
temp

**Review by Gambit142 — APPROVED** [review]
Hi @faisalakb @Omwomo  👋,

## Status: Approved 🟢   

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉

## Highlights
- Setup linters correctly ✔️ 
- Great use of `cancancan` gem ✔️ 
- Created `Ability` class that defines roles for users ✔️ 

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

#### Please tag me in your question so I can receive the notification. You can tag me by typing @Gambit142

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**